compiler/z/codegen/OMRMemoryReference.cpp

```diff
@@ -1407,11 +1409,37 @@ OMR::Z::MemoryReference::populateAddTree(TR::Node * subTree, TR::CodeGenerator *
 
    if ((integerChild->getOpCodeValue() == TR::isub || integerChild->getOpCodeValue() == TR::lsub))
       {
-      if (integerChild->getRegister() == NULL && _indexRegister->isLive())
-         {
-         TR_LiveRegisterInfo * liveRegister = _indexRegister->getLiveRegisterInfo();
+      /*
+       * arraycopy
+       *   aladd
+       *     ==>aRegLoad
+       *     lsub
+       *       ==>lRegLoad
+       *       lconst
+       *   aladd
+       *     ==>aconst NULL
+       *     lsub
+       *       ==>lRegLoad
+       *       ==>lconst
+       *
+       * With this tree, the first and second aladd get handled slightly differently.
+       * The first has a non NULL addressChild so that get's set as the _baseRegister.
+       * We will fold the lsub into the aladd's memref so the lRegLoad becomes the _indexRegister
+       * and the lconst becomes the offset.
+       * For the second aladd, the NULL addressChild means we use the integerChild as the _baseRegister.
+       * This is needed to prevent Zero Address Detection events on z/OS which happen when the base register of an instruction is zero.
+       * When we fold the lsub the lRegLoad becomes the _baseRegister and the lconst becomes offset.
+       * The important point is that in this case the _indexRegister is NULL.
+       * Thus we need to take care to select the appropriate register in the code below.
+       */
+      
+      TR::Register *integerChildRegister = _indexRegister ? _indexRegister : _baseRegister;
```

Comment:
Minor nit.  Can we make `(_indexRegister != NULL) ? ....` explicit here to improve readability?

Reply:
Changed.